diff --git a/src/any.rs b/src/any.rs
index 636cb92..84c64d9 100644
--- a/src/any.rs
+++ b/src/any.rs
@@ -12,7 +12,7 @@ pub trait CloneToAny {
 impl<T: Any + Clone> CloneToAny for T {
     #[inline]
     fn clone_to_any(&self) -> Box<dyn CloneAny> {
-        Box::new(self.clone())
+        Box::try_new(self.clone()).unwrap()
     }
 }
 
@@ -121,7 +121,7 @@ macro_rules! implement {
         impl<T: $any_trait $(+ $auto_traits)*> IntoBox<dyn $any_trait $(+ $auto_traits)*> for T {
             #[inline]
             fn into_box(self) -> Box<dyn $any_trait $(+ $auto_traits)*> {
-                Box::new(self)
+                Box::try_new(self).unwrap()
             }
         }
     }
diff --git a/src/lib.rs b/src/lib.rs
index 9c57f09..0ffdd30 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -21,6 +21,8 @@
 
 #![cfg_attr(not(feature = "std"), no_std)]
 
+#![feature(allocator_api)]
+
 use core::convert::TryInto;
 use core::hash::Hasher;
 
